### PR TITLE
feat: handle MPT transition block

### PR DESCRIPTION
**Pull request by Pangssu**

### CODE REVIEW

File: kroma-chain-ops/genesis/genesis.go [hunk 1]
```diff
@@ -62,8 +62,7 @@ func NewL2Genesis(config *DeployConfig, block *types.Block) (*core.Genesis, erro
 		ShanghaiTime:                  config.CanyonTime(block.Time()),
 		CancunTime:                    config.EcotoneTime(block.Time()),
 		EcotoneTime:                   config.EcotoneTime(block.Time()),
-		// TODO(seolaoh): uncomment this when geth updated
-		// KromaMPTTime:                  config.KromaMPTTime(block.Time()),
+		KromaMPTTime:                  config.KromaMPTTime(block.Time()),
 		InteropTime: config.InteropTime(block.Time()),
 		Kroma: &params.KromaConfig{
 			EIP1559Denominator:       eip1559Denom,
```
Comment (sm-stack) [inline kroma-chain-ops/genesis/genesis.go:68]: Shouldn't we add `Optimism` here?

File: op-node/rollup/superchain.go [hunk 2]
```diff
@@ -101,8 +101,6 @@ func LoadOPStackRollupConfig(chainID uint64) (*Config, error) {
 		CanyonTime:        chConfig.CanyonTime,
 		DeltaTime:         chConfig.DeltaTime,
 		EcotoneTime:       chConfig.EcotoneTime,
-		// TODO(seolaoh): uncomment this when geth updated
-		// KromaMPTTime:           chConfig.KromaMPTTime,
```
Comment (sm-stack) [inline op-node/rollup/superchain.go:105]: Why is this removed? Since setting the genesis block is irrelevant with KromaMPTTime?

File: kroma-chain-ops/genesis/genesis.go [hunk 3]
```diff
@@ -62,8 +62,7 @@ func NewL2Genesis(config *DeployConfig, block *types.Block) (*core.Genesis, erro
 		ShanghaiTime:                  config.CanyonTime(block.Time()),
 		CancunTime:                    config.EcotoneTime(block.Time()),
 		EcotoneTime:                   config.EcotoneTime(block.Time()),
-		// TODO(seolaoh): uncomment this when geth updated
-		// KromaMPTTime:                  config.KromaMPTTime(block.Time()),
+		KromaMPTTime:                  config.KromaMPTTime(block.Time()),
```
Comment (sm-stack) [inline kroma-chain-ops/genesis/genesis.go:65]: How about changing this to `KromaMPTBlock`? It will induce changes in geth, but I think it will be more coherent with the handling of Bedrock fork.

File: op-e2e/migration_test.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+	l1Cl := sys.Clients["l1"]
+	l2Seq := sys.Clients["sequencer"]
+	l2Verif := sys.Clients["verifier"]
+
+	// Send L2 TX
+	_, err = wait.ForTransferTxOnL2(sys.Cfg.L2ChainIDBig(), l2Seq, l2Verif, cfg.Secrets.Alice, common.Address{0xff, 0xff}, common.Big1)
+	require.NoError(t, err)
+
+	// Deploy contract on L2
+	minimalCode := common.Hex2Bytes("601680600c6000396000f3fe3615600b576000803555005b60016000540160005500")
```
[the remark this answers is not in the document]
Comment (Pangssu) [reply op-e2e/migration_test.go:106]: It simply writes a value to storage

File: kroma-chain-ops/genesis/genesis.go (hunk 3 above)
Comment (Pangssu) [reply kroma-chain-ops/genesis/genesis.go:65]: We've been doing upgrades by timestamp, not block number, so i thought it was better to make sure there was no confusion.

File: op-node/rollup/superchain.go (hunk 2 above)
Comment (Pangssu) [reply op-node/rollup/superchain.go:105]: Because it is not declared in the superchain-registry. Furthermore, we don't use this function.

**Review by kangsorang — APPROVED** [review]

File: op-node/rollup/derive/engine_controller.go [hunk 5]
```diff
@@ -327,7 +327,9 @@ func (e *EngineController) InsertUnsafePayload(ctx context.Context, envelope *et
 	// Check if there is a finalized head once when doing EL sync. If so, transition to CL sync
 	if e.syncStatus == syncStatusWillStartEL {
 		b, err := e.engine.L2BlockRefByLabel(ctx, eth.Finalized)
-		isTransitionBlock := e.rollupCfg.Genesis.L2.Number != 0 && b.Hash == e.rollupCfg.Genesis.L2.Hash
+		// [Kroma: ZKT to MPT]
+		isTransitionBlock := e.rollupCfg.Genesis.L2.Number != 0 && b.Hash == e.rollupCfg.Genesis.L2.Hash || e.rollupCfg.IsKromaMPT(ref.Time)
```
Comment (seolaoh) [inline op-node/rollup/derive/engine_controller.go:331]: I wonder why starting EL sync if the unsafe payload is after Kroma MPT time

File: ops-devnet/docker-compose.yml [hunk 6]
```diff
@@ -107,7 +107,7 @@ services:
 
   l2:
     pid: host # allow debugging
-    image: kromanetwork/geth:v0.5.0
+    image: kromanetwork/geth:mpt-d6420143
```
Comment (seolaoh) [inline ops-devnet/docker-compose.yml:110]: Which PR will you setup Kroma MPT devnet?

File: op-e2e/migration_test.go [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+	// Set historical rpc endpoint.
+	for name := range cfg.Nodes {
+		name := name
+		cfg.GethOptions[name] = append(cfg.GethOptions[name], []geth.GethOption{
+			func(ethCfg *ethconfig.Config, nodeCfg *node.Config) error {
+				// Since the migration process requires preimages, enable storing preimage option.
+				ethCfg.Preimages = true
+				ethCfg.RollupHistoricalRPC = fmt.Sprintf("http://127.0.0.1:%d", historicalRpcPort)
+				if name == "historical" {
```
[the remark this answers is not in the document]
Comment (seolaoh) [reply op-e2e/migration_test.go:77]: One question, so if a node wants to migrate, should the node be turned on with `--cache.preimages` flag?

[the remark this answers is not in the document]
Comment (Pangssu) [reply op-e2e/migration_test.go:77]: They can set `gcmode=archive` or `--cache.preimages=true` to proceed with the migration.

File: op-e2e/migration_test.go [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+					ethCfg.DisableMPTMigration = true
+				}
+				// Deep copy the genesis
+				dst := &core.Genesis{}
+				b, _ := json.Marshal(ethCfg.Genesis)
+				err := json.Unmarshal(b, dst)
+				if err != nil {
+					return err
+				}
+				ethCfg.Genesis = dst
```
[the remark this answers is not in the document]
Comment (Pangssu) [reply op-e2e/migration_test.go:88]: In e2e, when there are multiple nodes, if one node finalizes transition to mpt first, the shared genesis causes other nodes to also have their genesis changed. This is to avoid such an issue.

File: ops-devnet/docker-compose.yml (hunk 6 above)
Comment (Pangssu) [reply ops-devnet/docker-compose.yml:110]: I will change it after kromanetwork/go-ethereum#121 merged

File: op-e2e/migration_test.go [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+	"github.com/kroma-network/zktrie/trie"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils/geth"
+	rollupNode "github.com/ethereum-optimism/optimism/op-node/node"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/driver"
+	"github.com/ethereum-optimism/optimism/op-node/rollup/sync"
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/kroma-network/kroma/kroma-bindings/predeploys"
+	"github.com/kroma-network/kroma/op-e2e/e2eutils/wait"
```
[the remark this answers is not in the document]
Comment (Pangssu) [reply op-e2e/migration_test.go:31]: I think this is the right position because kroma-network/zktre is external module.

File: kroma-chain-ops/genesis/genesis.go (hunk 1 above)
Comment (Pangssu) [reply kroma-chain-ops/genesis/genesis.go:68]: added at L76

**Review by sm-stack — APPROVED** [review]
LGTM

File: op-node/rollup/derive/engine_controller.go (hunk 5 above)
Comment (Pangssu) [reply op-node/rollup/derive/engine_controller.go:331]: I checked again, it seems we don’t need to make any changes to this part.

File: ops-devnet/docker-compose.yml (hunk 6 above)
Comment (seolaoh) [reply ops-devnet/docker-compose.yml]: To setup Kroma MPT devnet, we need to add flag "--KromaZKTrie=false" to l2, add `l2GenesisKromaMPTTimeOffset` to devnet-template.json, and spin up historical node also (in docker-compose.yml, __init__.py)

Comment (Pangssu) [reply ops-devnet/docker-compose.yml]: I believe the issue has been resolved in https://github.com/kroma-network/go-ethereum/pull/119, so the `--KromaZKTrie=false` option is no longer necessary.

And is there a need to spin up historical node on the devnet? Could it be for testing purposes?

File: kroma-devnet/devnet/__init__.py [hunk 10]
```diff
@@ -256,6 +256,9 @@ def devnet_deploy(paths):
     run_command(['docker', 'compose', 'up', '-d', 'l2'], cwd=paths.ops_bedrock_dir, env={
         'PWD': paths.ops_bedrock_dir
     })
+    run_command(['docker', 'compose', 'up', '-d', 'l2-historical'], cwd=paths.ops_bedrock_dir, env={
+        'PWD': paths.ops_bedrock_dir
+    })
 
     # Wait for the L2 to be available.
```
[the remark this answers is not in the document]
Comment (Pangssu) [reply kroma-devnet/devnet/__init__.py:263]: added

**Review by seolaoh — APPROVED** [review]
